proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+Below are some examples:
+
+| Code | Diagnostic |
+|-|-|
+| `float f : register(t0)` | "error: 'float' is an invalid resource type for register type 't'" |
+| `float f : register(c0)` | Valid, no errors
+| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding. Disable with --Wno-disallow-legacy-binding-rules" |
+| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used. Disable with --Wno-disallow-legacy-binding-rules" |
+| `float f : register(x0)` | "error: register binding type 'x' not supported for variable of type 'float'" |
+| `cbuffer g_cbuffer { float f : register(c2); }` | "warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules" |
+| `struct Foo { RWBuffer<float> f; }; Foo foo : register(c2);`| "warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules" |
+| `RWBuffer<float> f : register(c3);`| "error: invalid register specification, expected 'u' binding" |
+
+
+## Detailed design
+
+All the compiler has to work with is a Decl object and the context the decl appears in.
+From this information, the first goal is to set a specific set of flags that can fully inform
+the right diagnostic to emit. The first group of flags are the decl type flags, which
+are either `basic`, `resource`, `udt`, or `other`. `basic` refers to a numeric variable
+type. `other` refers to a type that cannot possibly be resource, because it lacks the 
+resource attribute.<br>
+The next group of flags are the resource class flags,
+which are `srv`, `uav`, `cbv`, or `sampler`.<br>
+The final flag is `default_globals`, which indicates whether or not the decl appears
+inside the $Globals scope.<br><br>
+
+From the Decl object, first determine if the decl is a cbuffer or tbuffer decl. If so,
+we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
+respectively). Otherwise, the decl is classified according to the underlying canonical 
+type.<br>
+If the decl is implicit, then check if the decl is a vector, matrix, or otherwise numeric type,
+and if so, we know that the decl is a `basic` type. If the decl has a resource attribute, 
+we can set the `resource` flag, and can also obtain the resource class and set the resource 
+class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
+the `other` flag is set.<br>
+If the decl is not implicit, then we check if it is a struct or class type, and if so, the
+decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
+collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
+sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
+type, then it is a basic type, and so the `basic` flag is set.<br>

```

Comment:
```suggestion
If the decl type is not implicit, then we check if it is a struct or class type.
If so, the decl is a UDT, so the `udt` flag is set, otherwise, it's a basic type,
so we set the `basic` flag.

For a `udt`, recurse through the members of the UDT and its bases,
collecting information on what types of resources are contained by the UDT,
setting corresponding class flags (`uav`, `cbv`, `srv`, or `sampler`).
Also track whether there are any other numeric members and set a
`numeric` flag.

```